Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPipelineState.h

```diff
@@ -0,0 +1,413 @@
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#ifndef plMetalPipelineState_h
+#define plMetalPipelineState_h
+
+#include <stdio.h>
+
+#include <Metal/Metal.hpp>
+
+#include "plMetalDevice.h"
+#include "plSurface/plShaderTable.h"
+
+//MARK: Base pipeline state
+
+class plMetalPipelineState
+{
+public:
+    plMetalPipelineState(plMetalDevice* device);
+    plMetalDevice::plMetalLinkedPipeline* GetRenderPipelineState();
+    void                                  PrewarmRenderPipelineState();
+    bool                                  operator==(const plMetalPipelineState& p) const
+    {
+        if ((&p)->GetID() != this->GetID()) {
+            return false;
+        } else {
+            return IsEqual(p);
+        }
+    }
+    virtual size_t                GetHash() const;
+    virtual bool                  IsEqual(const plMetalPipelineState& p) const = 0;
+    virtual uint16_t              GetID() const { return 0; };
+    virtual plMetalPipelineState* Clone() = 0;
+
+    //
+    virtual const MTL::Function* GetVertexFunction(MTL::Library* library) = 0;
+    virtual const MTL::Function* GetFragmentFunction(MTL::Library* library) = 0;
+    virtual const NS::String*    GetDescription() = 0;
+
+    virtual void ConfigureBlend(MTL::RenderPipelineColorAttachmentDescriptor* descriptor) = 0;
+    virtual void ConfigureVertexDescriptor(MTL::VertexDescriptor* vertexDescriptor) = 0;
+    virtual ~plMetalPipelineState() = default;
```

Comment:
Let's move this to be next to the ctor. I almost didn't see it and left a comment about needing a virtual dtor.